coderd/database/dbtestutil/postgres.go

```diff
@@ -16,119 +25,452 @@ import (
+)
+
+// initDefaultConnectionParams initializes the default postgres connection parameters.
+// It first checks if the database is running at localhost:5432. If it is, it will
+// use that database. If it's not, it will start a new container and use that.
+func initDefaultConnectionParams() error {
+	params := ConnectionParams{
+		Username: "postgres",
+		Password: "postgres",
+		Host:     "127.0.0.1",
+		Port:     "5432",
+		DBName:   "postgres",
+	}
+	dsn := params.DSN()
+	db, err := sql.Open("postgres", dsn)
+	if err == nil {
+		err = db.Ping()
+		if closeErr := db.Close(); closeErr != nil {
+			return xerrors.Errorf("close db: %w", closeErr)
+		}
+	}
+	shouldOpenContainer := false
+	if err != nil {
+		errSubstrings := []string{
+			"connection refused",          // this happens on Linux when there's nothing listening on the port
+			"No connection could be made", // like above but Windows
+		}
+		errString := err.Error()
+		for _, errSubstring := range errSubstrings {
+			if strings.Contains(errString, errSubstring) {
+				shouldOpenContainer = true
+				break
+			}
+		}
+	}
+	if err != nil && shouldOpenContainer {
+		// If there's no database running on the default port, we'll start a
+		// postgres container. We won't be cleaning it up so it can be reused
+		// by subsequent tests. It'll keep on running until the user terminates
+		// it manually.
+		container, _, err := openContainer(DBContainerOptions{
+			Name: "coder-test-postgres",
+			Port: 5432,
+		})
 		if err != nil {
-			return "", nil, xerrors.Errorf("connect to ci postgres: %w", err)
+			return xerrors.Errorf("open container: %w", err)
 		}
+		params.Host = container.Host
+		params.Port = container.Port
+		dsn = params.DSN()
 
-		defer db.Close()
+		// Retry connecting for a cumulative 5 seconds.
+		// The fact that openContainer succeeded does not
+		// mean that port forwarding is ready.
+		for i := 0; i < 20; i++ {
+			db, err = sql.Open("postgres", dsn)
+			if err == nil {
+				err = db.Ping()
+				if closeErr := db.Close(); closeErr != nil {
+					return xerrors.Errorf("close db, container: %w", closeErr)
+				}
+			}
+			if err == nil {
+				break
+			}
+			time.Sleep(250 * time.Millisecond)
+		}
+	} else if err != nil {
+		return xerrors.Errorf("open postgres connection: %w", err)
+	}
+	defaultConnectionParams = params
+	return nil
+}
+
+type OpenOptions struct {
+	DBFrom *string
+}
+
+type OpenOption func(*OpenOptions)
+
+// WithDBFrom sets the template database to use when creating a new database.
+// Overrides the DB_FROM environment variable.
+func WithDBFrom(dbFrom string) OpenOption {
+	return func(o *OpenOptions) {
+		o.DBFrom = &dbFrom
+	}
+}
 
-		dbName, err := cryptorand.StringCharset(cryptorand.Lower, 10)
+// Open creates a new PostgreSQL database instance.
+// If there's a database running at localhost:5432, it will use that.
+// Otherwise, it will start a new postgres container.
+func Open(opts ...OpenOption) (string, func(), error) {
```

Comment:
Should this be `Open(t *testing.T, ...)` and not return a `cleanup` func?

Then we do `t.Cleanup(close())` inside this? I know it is more refactoring :cry: